A1.md

```diff
@@ -0,0 +1,133 @@
+     used for.
+
+We propose the following functionality to support these two cases:
+
+- For case 1, the client channel code will check for the existence of the
+  `http_proxy` environment variable (or equivalent signal).  If that
+  signal is present, the client channel code will do the following:
+  - Ask the resolver to resolve the proxy name instead of the server name
+    requested via the client API.  (Note that this will cause us to
+    create a single subchannel to the proxy, as opposed to one
+    subchannel for each backend server.)
+  - Add state (e.g., a channel argument in C-core) requesting that the
+    subchannel code (see below) send an HTTP CONNECT request for the
+    server name requested via the client API.
+
+- For case 2, in the subchannel, before we connect to the target address,
+  we will call a new *proxy mapper* hook, which will allow selectively
+  requesting the use of a proxy based on the address to which the
+  subchannel is going to connect.  If a proxy is to be used, the mapper
+  will return the proxy address (which the subchannel will connect to
+  instead of the address returned by the resolver) and some additional
+  state (e.g., a channel argument in C-core) requesting that the
+  subchannel send an HTTP CONNECT request for the original subchannel
+  address.  Note that in this case, the server name used in the HTTP
+  CONNECT request will be an IP address and port, not a hostname (i.e.,
+  we are not relying on the proxy to do DNS resolution for us).
+
+- When the subchannel establishes a new connection to a proxy, it will
+  optionally (based on the state added in one of the two steps above)
+  send an HTTP CONNECT request indicating the server it needs to connect to.
+
+Note that in case 1, because the client cannot know the set of server
+addresses, it is impossible to use the normal gRPC per-call load balancing.
+It *is* possible to do load balancing on a per-connection basis, but
+that may not spread out the load evenly if different clients impose a
+different amount of load.
+
+In contrast, in case 2, it *is* possible to do the normal gRPC per-call
+load balancing, because the client does know the set of server
+addresses.  The fact that the connections to those individual addresses
+go through a TCP-level proxy does not interfere with the client opening
+a separate subchannel to each address.
+
+Also, note that in case 1, the client will not have access to the
+service config for the server name.  This is not a problem in case 2.
+
+## Rationale
+
+The proposed design allows two different ways of triggering use of the
+HTTP CONNECT handshaker code, which addresses the needs of both
+environments described above.
+
+There have also been discussions about how to do proxying for an
+environment where a service is accessible over the internet but its
+internal architecture (e.g., DNS names or server addresses) should not
+be exposed publicly.  In this case, it is impossible to support per-call
```

Comment:
`impossible` is too strong a word. See above comment about the role of the proxy mapper with respect to the HTTP CONNECT request.